### ISSUE TITLE
Verify response content length in IISHttpContextOfT

### ISSUE THREAD

**Issue by jkotalik** [aspnet/IISIntegration#433]
Code is currently commented out. Based on https://github.com/aspnet/KestrelHttpServer/blob/a31d1e024ca86a4a9053e0be105fde17c5ccae10/src/Kestrel.Core/Internal/Http/HttpProtocol.cs#L454.

**Comment by jkotalik** [comment]
@muratg Lower priority, but something that needs to happen before 2.1 release.

**Comment by davidfowl** [comment]
We should verify that's IIS doesn't already do this.

**Comment by muratg** [comment]
@jkotalik please check if IIS support this.

**Comment by muratg** [comment]
HTTP.SYS already provides this. No need to do it on ANCM level.

**Comment by muratg** [comment]
@jhkimnew Could you verify what the behavior here is? @pan-wang has some concerns about the case where app sets a content length header.

**Comment by Tratcher** [comment]
HTTP.SYS doesn't provide this. Does IIS? Is an error surface to the application code when they've written too many or too few bytes?

**Comment by jhkimnew** [comment]
I checked the current IIS code how it handles content-length. Here is a simplified algorithm.

1. If there is a single response (no more data to send), IIS simply the content-length header if it is not HEAD request. That means http.sys should set content-length.
2. If the response is not a single response (more data left to complete the response), IIS does not delete header trusting application set the content-length correctly.

Releated IIS functions are listed below so that you can refer to.

iiscore!W3_RESPONSE::DoActualSendResponse
iiscore!W3_RESPONSE::ResumeResponseTransfer
iiscore!W3_RESPONSE::CalculateAndAddContentLength

**Comment by davidfowl** [comment]
It's interesting that IIS doesn't have this check. I wonder why... Also I wonder if ASP.NET handled this at all.

**Comment by muratg** [comment]
Bringing back to triage...
